### PR TITLE
Initial changes for the new camera rig configuration

### PR DESCRIPTION
**Pull request by StephenHodgson**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

After updating the solvers in the last PR I did, I realized that I didn't actually make the solvers body locked but playspace locked. This PR aims to correct the body locking features by extending the `IMixedRealityCameraSystem` to include options for a `IMixedRealityCameraRig` that can be implemented on a `MonoBehavior` and used to reference the camera rig components and transforms.

## Target of the change:

Is this enhancement for:

- Core (core framework, interfaces and definitions)

## Changes:

Brief list of the targeted features that are being changed.

- Fixes: Body Locked/Playspace Locked behavior in the solver system.

## Breaking Changes:

- Added a few new component types to the `IMixedRealityCameraSystem`:
  - `IMixedRealityCameraRig` which keeps track of the playspace, camera, and body transforms.
  - `MixedRealityCameraProfile` now has two additional configuration options:
    - `CameraRigType` to specify the concrete type of `IMixedRealityCameraRig` to use
    - `bodyAdjustmentAngle` to turn the player's body when the angle between the head and body has reached this threshold.
- Updated `TrackedObjectType` to include `Playspace`
- Updated `SolverHandler` to properly setup and track the `Body` and `Playspace` tracked objects

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
I like the overall concept of the camera rig and it's overall space and relationship to the camera/playspace, a lot more coherent than just a single playspace reference.

**Comment by StephenHodgson** [comment]
Okay guys this is ready for review

**Comment by StephenHodgson** [comment]
Scratch that. I'm getting a weird error with the gaze pointer.

```
Unable to find IMixedRealityInputSystem service.
	UnityEngine.Debug:LogError(Object)
	XRTK.Services.MixedRealityToolkit:GetService(Type, String, Boolean) (at Packages/com.xrtk.core/Services/MixedRealityToolkit.cs:1553)
	XRTK.Services.MixedRealityToolkit:GetService(Type, Boolean) (at Packages/com.xrtk.core/Services/MixedRealityToolkit.cs:1539)
	XRTK.Services.MixedRealityToolkit:GetService(Boolean) (at Packages/com.xrtk.core/Services/MixedRealityToolkit.cs:1488)
	XRTK.Services.MixedRealityToolkit:get_InputSystem() (at Packages/com.xrtk.core/Services/MixedRealityToolkit.cs:1758)
	XRTK.SDK.Input.<>c:<.ctor>b__5_0() (at Packages/com.xrtk.sdk/Features/Input/InputSystemGlobalListener.cs:17)
	UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
	XRTK.Utilities.Async.AwaiterExtensions:RunCoroutine(IEnumerator) (at Packages/com.xrtk.core/Utilities/Async/AwaiterExtensions.cs:150)
	XRTK.Utilities.Async.<>c__DisplayClass14_0:<GetAwaiterReturnVoid>b__0() (at Packages/com.xrtk.core/Utilities/Async/AwaiterExtensions.cs:136)
	XRTK.Utilities.Async.AwaiterExtensions:RunOnUnityScheduler(Action) (at Packages/com.xrtk.core/Utilities/Async/AwaiterExtensions.cs:157)
	XRTK.Utilities.Async.AwaiterExtensions:GetAwaiterReturnVoid(Object) (at Packages/com.xrtk.core/Utilities/Async/AwaiterExtensions.cs:136)
	XRTK.Utilities.Async.AwaiterExtensions:GetAwaiter(WaitUntil) (at Packages/com.xrtk.core/Utilities/Async/AwaiterExtensions.cs:79)
	XRTK.SDK.Input.<Start>d__3:MoveNext() (at Packages/com.xrtk.sdk/Features/Input/InputSystemGlobalListener.cs:31)
	System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start(<Start>d__3&)
	XRTK.SDK.Input.InputSystemGlobalListener:Start()
	XRTK.SDK.Input.GazeProvider:<>n__0()
	XRTK.SDK.Input.<Start>d__55:MoveNext() (at Packages/com.xrtk.sdk/Features/Input/GazeProvider.cs:259)
	System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start(<Start>d__55&)
	XRTK.SDK.Input.GazeProvider:Start()
	
NullReferenceException: Object reference not set to an instance of an object
	XRTK.Services.InputSystem.Pointers.GenericPointer..ctor (System.String pointerName, XRTK.Interfaces.InputSystem.IMixedRealityInputSource inputSourceParent) (at Packages/com.xrtk.core/Services/InputSystem/GenericPointer.cs:27)
	XRTK.SDK.Input.GazeProvider+InternalGazePointer..ctor (XRTK.SDK.Input.GazeProvider gazeProvider, System.String pointerName, XRTK.Interfaces.InputSystem.IMixedRealityInputSource inputSourceParent, UnityEngine.LayerMask[] raycastLayerMasks, System.Single pointerExtent, UnityEngine.Transform gazeTransform, XRTK.Utilities.Physics.BaseRayStabilizer stabilizer) (at Packages/com.xrtk.sdk/Features/Input/GazeProvider.cs:146)
	XRTK.SDK.Input.GazeProvider.InitializeGazePointer () (at Packages/com.xrtk.sdk/Features/Input/GazeProvider.cs:369)
	XRTK.SDK.Input.GazeProvider.get_GazePointer () (at Packages/com.xrtk.sdk/Features/Input/GazeProvider.cs:98)
	XRTK.SDK.Input.GazeProvider.get_GazeOrigin () (at Packages/com.xrtk.sdk/Features/Input/GazeProvider.cs:117)
	XRTK.SDK.Input.GazeProvider.LateUpdate () (at Packages/com.xrtk.sdk/Features/Input/GazeProvider.cs:283)
	

```

**Comment by StephenHodgson** [comment]
Okay really ready now. I had to fix a few things

File: XRTK-Core/Packages/com.xrtk.core/Interfaces/IMixedRealityCameraRig.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        /// </summary>
+        Transform CameraTransform { get; }
+
+        /// <summary>
+        /// The player's <see cref="Camera"/> reference.
+        /// </summary>
+        Camera PlayerCamera { get; }
+
+        /// <summary>
+        /// The player's body transform. This <see cref="Transform"/> is synced to the player's head X & Z values
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Interfaces/IMixedRealityCameraRig.cs]:
```suggestion
        /// The player's body transform. This <see cref="Transform"/> is synced to the player's head X &amp; Z values
```

**Comment by SimonDarksideJ** [comment]
There seems to be an issue with teleportation related to this change.  If I put down a plane, then teleport to a corner, you don't end up in the corner, you are positioned a distance away from where you selected.

**Comment by StephenHodgson** [comment]
Okay, I'll be sure to check that out.

File: XRTK-Core/Packages/com.xrtk.core/Services/TeleportSystem/MixedRealityTeleportSystem.cs [hunk 2]
```diff
@@ -220,7 +220,7 @@ private void ProcessTeleportationRequest(TeleportEventData eventData)
         {
             isProcessingTeleportRequest = true;
 
-            var cameraParent = MixedRealityToolkit.Instance.MixedRealityPlayspace;
+            var cameraParent = MixedRealityToolkit.CameraSystem.CameraRig.PlayspaceTransform;
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/TeleportSystem/MixedRealityTeleportSystem.cs]: @SimonDarksideJ This is the only change to the teleport system.

Maybe we need to update this to be the playspace

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Services/TeleportSystem/MixedRealityTeleportSystem.cs]: I'll go back to the development branch and compare in the morning, Will do a couple of GIF's and see what's going on

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Services/TeleportSystem/MixedRealityTeleportSystem.cs]: Were you getting the same behavior on the dev branch?

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Services/TeleportSystem/MixedRealityTeleportSystem.cs]: Tested against release and dev and found this change is not causing the teleportation issue.  Raised #390

**Pull request merged**